cpp/src/io/parquet/experimental/hybrid_scan_helpers.cpp

```diff
@@ -249,40 +264,238 @@ aggregate_reader_metadata::select_payload_columns(
+                    // For all column chunks
+                    std::for_each(
+                      bloom_filter_col_schemas.begin(),
+                      bloom_filter_col_schemas.end(),
+                      [&](auto const schema_idx) {
+                        auto& col_meta = get_column_metadata(rg_index, src_index, schema_idx);
+                        // Get bloom filter offsets and sizes
+                        bloom_filter_bytes.emplace_back(col_meta.bloom_filter_offset.value_or(0),
+                                                        col_meta.bloom_filter_length.value_or(0));
+
+                        // Set `have_bloom_filters` if `bloom_filter_offset` is valid
+                        if (col_meta.bloom_filter_offset.has_value()) { have_bloom_filters = true; }
+                      });
+                  });
+                });
+
+  if (not have_bloom_filters) { return {}; }
+
+  return bloom_filter_bytes;
+}
+
+std::vector<byte_range_info> aggregate_reader_metadata::get_dictionary_page_bytes(
+  cudf::host_span<std::vector<cudf::size_type> const> row_group_indices,
+  host_span<data_type const> output_dtypes,
+  host_span<cudf::size_type const> output_column_schemas,
+  std::reference_wrapper<ast::expression const> filter)
+{
+  // Collect (in)equality literals for each input table column
+  auto const literals =
+    dictionary_literals_collector{filter.get(), static_cast<cudf::size_type>(output_dtypes.size())}
+      .get_literals();
+
+  auto iter = thrust::make_zip_iterator(thrust::counting_iterator<cudf::size_type>(0),
+                                        output_column_schemas.begin());
+
+  // Collect schema indices of columns with equality predicate(s)
+  std::vector<thrust::tuple<cudf::size_type, cudf::size_type>> dictionary_col_schemas;
+  thrust::copy_if(thrust::host,
+                  iter,
+                  iter + output_column_schemas.size(),
+                  literals.begin(),
+                  std::back_inserter(dictionary_col_schemas),
+                  [](auto& dict_literals) { return not dict_literals.empty(); });
+
+  // No (in)equality literals found, return empty vector
+  if (dictionary_col_schemas.empty()) { return {}; }
+
+  // Compute total number of input row groups
+  auto const total_row_groups = compute_total_row_groups(row_group_indices);
+
+  // Descriptors for all the chunks that make up the selected columns
+  auto const num_dictionary_columns = dictionary_col_schemas.size();
+  auto const num_chunks             = total_row_groups * num_dictionary_columns;
+
+  std::vector<byte_range_info> dictionary_page_bytes;
+  dictionary_page_bytes.reserve(num_chunks);
+
+  // Flag to check if we have at least one valid dictionary page
+  auto have_dictionary_pages = false;
+
+  // For all sources
+  std::for_each(
+    thrust::counting_iterator<size_t>(0),
+    thrust::counting_iterator(row_group_indices.size()),
```

Comment:
This loop is called multiple times. Can we extract it to a common function?

Reply:
The innermost loop bounds and contents are different in both locations of this loop. I had them consolidated in an earlier version of this PR but that makes the code unreadable.